tmc-langs-qmake/src/main/java/fi/helsinki/cs/tmc/langs/qmake/QmakePlugin.java

```diff
@@ -0,0 +1,311 @@
+import fi.helsinki.cs.tmc.langs.domain.ExerciseDesc;
+import fi.helsinki.cs.tmc.langs.domain.RunResult;
+import fi.helsinki.cs.tmc.langs.domain.RunResult.Status;
+import fi.helsinki.cs.tmc.langs.domain.SpecialLogs;
+import fi.helsinki.cs.tmc.langs.domain.TestDesc;
+import fi.helsinki.cs.tmc.langs.domain.TestResult;
+import fi.helsinki.cs.tmc.langs.io.StudentFilePolicy;
+import fi.helsinki.cs.tmc.langs.io.sandbox.StudentFileAwareSubmissionProcessor;
+import fi.helsinki.cs.tmc.langs.io.zip.StudentFileAwareUnzipper;
+import fi.helsinki.cs.tmc.langs.io.zip.StudentFileAwareZipper;
+import fi.helsinki.cs.tmc.langs.utils.ProcessResult;
+import fi.helsinki.cs.tmc.langs.utils.ProcessRunner;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.regex.MatchResult;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class QmakePlugin extends AbstractLanguagePlugin {
+
+    private static final Path TMC_TEST_RESULTS = Paths.get("tmc_test_results.xml");
+
+    // POINT(exercise_name, 1)
+    private static final Pattern POINT_PATTERN
+            = Pattern.compile("POINT\\(\\s*(\\w+),\\s*(\\w+)\\s*\\)\\s*;");
+    private static final Pattern COMMENT_PATTERN
+            = Pattern.compile("(^[^\"\\r\\n]*\\/\\*{1,2}.*?\\*\\/|(^[^\"\\r\\n]*\\/\\/[^\\r\\n]*))", Pattern.MULTILINE | Pattern.DOTALL);
+
+    private static final RunResult EMPTY_FAILURE
+            = new RunResult(
+                    Status.COMPILE_FAILED,
+                    ImmutableList.<TestResult>of(),
+                    new ImmutableMap.Builder<String, byte[]>().build());
+
+    private static final Logger log = LoggerFactory.getLogger(QmakePlugin.class);
+
+    /**
+     * Creates a new QmakePlugin.
+     */
+    public QmakePlugin() {
+        super(
+                new ExerciseBuilder(),
+                new StudentFileAwareSubmissionProcessor(),
+                new StudentFileAwareZipper(),
+                new StudentFileAwareUnzipper());
+    }
+
+    @Override
+    public String getPluginName() {
+        return "qmake";
+    }
+
+    /**
+     * Resolve the exercise .pro file from exercise directory. The file should
+     * be named after the directory.
+     */
+    private Path getProFile(Path basePath) {
+        Path proFile = new File(basePath.toFile().getName() + ".pro").toPath();
```

Comment:
Thanks, did not know about the old/new java.io API!

Reply:
Most of the old `io` package is fine, but it has enough stupid "gotcha" stuff that it's just easier to avoid it wholesale :)

For example, if you have two `java.io.URL` objects, [`java.io.URL.equals()`](https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#equals(java.lang.Object)) ends up making two DNS calls and thinks the hosts are equal if they resolve to the same IP: something that might have been only a slight horror in the age of dedicated self-hosted servers, but is completely useless and actively harmful in the age of cloud computing.